ddl/cache.lua

```diff
@@ -0,0 +1,108 @@
+local fiber = require('fiber')
+
+local cache = nil
+
+local CACHE_LOCK_TIMEOUT  = 3
+local SPACE_NAME_IDX      = 1
+local SHARD_FUNC_NAME_IDX = 2
+local SHARD_FUNC_BODY_IDX = 3
+
+-- Build cache and setup "on_replace" trigger.
+--
+-- Cache structure format:
+-- cache = {
+--   spaces = {
+--     'space_name' = {
+--       raw = {},      -- raw sharding metadata, used for ddl.get()
+--       processed = {} -- sharding function that is ready to call or a string with error
+--     }
+--   },
+--   fetch_lock -- locking based on fiber.channel()
+-- }
+local function cache_build()
+    -- prevent cache_build() from being called concurrently by different fibers
+    local unlocked = cache.fetch_lock:put(true, CACHE_LOCK_TIMEOUT)
+    local trigger_found = false
+
+    if not unlocked then
+        return
+    end
```

Comment:
Something going wrong... shouldn't we return an error?

Reply:
fixed